runtime/src/main/cpp/CyclicCollector.cpp

```diff
@@ -0,0 +1,435 @@
+           auto* obj = toVisit.front();
+           toVisit.pop_front();
+           auto* objContainer = obj->container();
+           if (objContainer == nullptr) continue;  // Permanent object.
+           if (!objContainer->frozen()) continue;
+           sideRefCounts_[obj] = -1;
+           visited.insert(obj);
+           if (atomicGet(&mutatedAtomics_) != 0) {
+             COLLECTOR_LOG("restarted during reachable visit\n")
+             restartCount++;
+             goto restart;
+           }
+           traverseObjectFields(obj, [&toVisit, &visited](ObjHeader** location) {
+              ObjHeader* ref = *location;
+              if (ref != nullptr && (visited.count(ref) == 0)) {
+                toVisit.push_back(ref);
+              }
+           });
+         }
+         // Now release all atomic roots with matching reference counters, as only their destruction is controlled.
+         for (auto it: sideRefCounts_) {
+           auto* obj = it.first;
+           // Only do that for atomic rootset elements.
+           if ((obj->type_info()->flags_ & TF_LEAK_DETECTOR_CANDIDATE) == 0) {
+             continue;
+           }
+           auto* objContainer = obj->container();
+           if (!objContainer->frozen()) continue;
+           COLLECTOR_LOG("for %p inner %d actual %d\n", obj, it.second, objContainer->refCount());
+           // All references are inner. We compare the number of counted
+           // inner references with the number of non-stack references and per-thread ownership value
+           // (see rememberNewContainer()).
+           if (it.second == objContainer->refCount()) {
+             COLLECTOR_LOG("adding %p to release candidates\n", it.first);
+             toRelease_.insert(it.first);
+           }
+         }
+         if (toRelease_.size() > 0)
+           atomicSet(&pendingRelease_, 1);
+         atomicSet(&gcRunning_, 0);
+         shallRunCollector_ = false;
+       }
+     }
+     atomicSet(&terminateCollector_, false);
+  }
+
+  void addWorker(void* worker) {
+    Locker lock(&lock_);
+    currentAliveWorkers_++;
+    if (mainWorker_ == nullptr) mainWorker_ = worker;
+  }
+
+  void removeWorker(void* worker) {
+    Locker lock(&lock_);
+    // When exiting the worker - we shall collect the cyclic garbage here.
+    shallCollectGarbage_ = true;
+    shallRunCollector_ = true;
+    CHECK_CALL(pthread_cond_signal(&cond_), "Cannot signal collector")
+    currentAliveWorkers_--;
+  }
+
+  void addRoot(ObjHeader* obj) {
+    COLLECTOR_LOG("add root %p\n", obj);
+    // TODO: we can only add root when collector is not processing, which looks like a limitation,
+    //  instead we can add elements to the side buffer or have a separate lock for that.
+    Locker lock(&lock_);
+    rootset_.insert(obj);
+  }
+
+  void removeRoot(ObjHeader* obj) {
+    COLLECTOR_LOG("remove root %p\n", obj);
+    // Note that we can only remove root when the collector is not processing.
+    Locker lock(&lock_);
```

Comment:
Can this block the main thread?

Reply:
Yes, modifications to atomic rootset are blocking wrt GC lock, see no way of fixing it.

Reply:
Memory-wise, how big does the rootset data structure get, realistically?

Would it be possible to post all modifications as events in a non-blocking queue, and rebuild the state on the side, asynchronously, to avoid any locking whatsoever?

Reply:
Linear by size of all data rooted in atomic references. Not sure about feasibility of the proposal, it's somewhat too heavyweight to make the whole system transactional.

Reply:
But the only part we're interested in accessing lock-free are the _pointers/references_ to the data, not the data itself, no? Wouldn't that be a lot smaller? Unless those are not stable...?

Reply:
Well, you could probably try prototype to describe this idea, I'm not seeing how it could work so far.